tutor/templates/build/openedx/Dockerfile

```diff
@@ -152,14 +152,20 @@ RUN if [ "$APP_USER_ID" = 0 ]; then echo "app user may not be root" && false; fi
 RUN useradd --no-log-init --home-dir /openedx --create-home --shell /bin/bash --uid ${APP_USER_ID} app
 USER ${APP_USER_ID}
 
+# Note:
+# For directories from other stages, we prefer 'COPY --link' to plain 'COPY' because it copies
+# without regard to files from previous layers, providing significant caching benefits. However,
+# since Linux's username->userid mapping is stored in a file (/etc/passwd), it means that we must
+# --chown with an integer user id ($APP_USER_ID) rather the a username (app).
```

Comment:
Maybe I lack the context, but why do we need the ID here instead of username? Docker docs suggest we can use both username and numerical ID. looking at line 152, the user 'app' ID is APP_USER_ID.

Reply:
@DawoudSheraz We must use the ID because of the switch from `COPY` to `COPY --link`.

With `COPY`, the copy instruction is executed *using the state of the previous layer*. For example:

```Dockerfile
FROM ubuntu AS myimage

# This will create 'dawoud' in /etc/passwd, with user id == 3000
RUN useradd dawoud --uid 3000

# This will read /etc/password to determine that dawoud's user id is 3000, allowing it to set
# the owner of /myfile to 3000.
COPY --chown=dawoud /myfile /myfile
```

However, with `COPY --link`, the copy instruction is executed *independently* using the base image, and then merged together at the end. For example:
```Dockerfile
FROM ubuntu AS myimage

# This will create 'dawoud' in /etc/passwd, with user id == 3000
RUN useradd dawoud --uid 3000

# This will try to read /etc/password to determine dawoud's user id, but it be reading from
# ubuntu's default blank /etc/passwd, which will not contain 'dawoud'.
# An error will be raised and the build will fail.
COPY --link --chown=dawoud /myfile /myfile
```

The correct way to write this dockerfile would be:
```Dockerfile
FROM ubuntu AS myimage
ENV DAWOUD_USER_ID 3000
RUN useradd dawoud --uid $DAWOUD_USER_ID

# Even though dawoud is not in /etc/passwd at this point, COPY can still set user id to 3000
COPY --link --chown=$DAWOUD_USER_ID /myfile /myfile
```

In the final image, Docker will merge together the results of `RUN useradd` and `COPY --link`, creating a complete filesystem containing both `/etc/passwd` with 'dawoud', as well as `/myfile`.

Does that make sense @DawoudSheraz ? It took me a long time to figure this out and understand it myself 😓

Reply:
Thanks for the detailed explanation.